tools/xkbcli-scaffold-new-layout.py

```diff
@@ -0,0 +1,323 @@
+#  the rights to use, copy, modify, merge, publish, distribute, sublicense,
+#  and/or sell copies of the Software, and to permit persons to whom the
+#  Software is furnished to do so, subject to the following conditions:
+#
+#  The above copyright notice and this permission notice (including the next
+#  paragraph) shall be included in all copies or substantial portions of the
+#  Software.
+#
+#  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+#  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+#  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
+#  THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+#  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+#  FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+#  DEALINGS IN THE SOFTWARE.
+
+
+# This script creates the necessary scaffolding to create a custom keyboard
+# layout or option. It does not actually configure anything, it merely creates
+# the required directory structure and file scaffolding for the key
+# configurations to be added by the user.
+#
+
+import argparse
+import logging
+import os
+import re
+import sys
+
+from pathlib import Path
+from textwrap import dedent
+
+
+# Default values are set by meson but to make this usable directly within the
+# git source tree, use some sensible default values and return those where the
+# meson define hasn't been replaced.
+def default_value(key):
+    defaults = {
+        'extrapath': ('@XKBEXTRAPATH@', '/etc/xkb'),
+        'rules': ('@DEFAULT_XKB_RULES@', 'evdev'),
+    }
+
+    mesondefault, default = defaults[key]
+    if mesondefault.startswith('@') and mesondefault.endswith('@'):
+        return default
+    else:
+        return mesondefault
+
+
+logging.basicConfig(level=logging.DEBUG)
+logger = logging.getLogger('xkbcli')
+logger.setLevel(logging.INFO)
+
+
+def create_directory_structure(basedir):
+    basedir.mkdir(exist_ok=True)
+    # Note: we skip geometry
+    for d in ['compat', 'keycodes', 'rules', 'symbols', 'types']:
+        (basedir / d).mkdir(exist_ok=True)
+
+
+def create_rules_template(base_path, ruleset, layout, option):
+    rules = Path(base_path) / 'rules' / ruleset
+    if rules.exists():
+        logger.warning(f'Rules file {rules} already exists, skipping')
+        return
+
+    with open(rules, 'w') as fd:
+        header = dedent(f'''\
+            // generated by xkbcli scaffold-new-layout
+
+            // Note: no rules file entries are required for for a custom layout
+
+            ''')
+        fd.write(header)
+
+        if option:
+            group, section = option
+            option_template = dedent(f'''\
+            // This section maps XKB option "{group}:{section}" to the '{section}' section in the
+            // 'symbols/{group}' file.
+            //
+            ! option = symbols
```

Comment:
This assumed that the option is a symbols option, but options for compat/types/keycodes are possible as well. I agree `symbols` is probably the best default, but perhaps we can mention this in the generated comment above.

Reply:
I got curious and ran this command in `xkeyboard-config`:

```
$ git log --format=oneline --since 2015 keycodes compat types | wc -l
18
```
That's for an almost 5-year span. Quick glance shows only 11 of those are non-janitor commits and those are largely adding new keycodes in the form of, e.g.

```
+       <I372> = 372;   // #define KEY_FAVORITES           364
```
so yes, it's just a comment but at this point I'm wondering whether that would point users down a path they don't actually need to travel. The ones that care about compat and types will probably figure it out independently.

Completely unrelated: I do wonder if we should built automatic mapping for evdev keycodes into the parser so the above addition isn't needed.

Reply:
I've added a short blurb to the documentation itself for this.